src/katgpucbf/xbgpu/bsend.py

```diff
@@ -0,0 +1,390 @@
+    heaps_per_fengine_per_chunk
+        Number of SPEAD heaps from one F-engine in a single received Chunk.
+    n_tx_items
+        TODO: There must be a better way to say this
+        Number of Chunks to create in order to download data off the GPU.
+    n_channels_per_substream, spectra_per_heap, channel_offset
+        See :class:`.xbgpu.engine.XBEngine` for further information.
+    timestamp_step
+        The timestamp step between successive heaps, as dictated by the XBEngine.
+    send_rate_factor
+        Factor dictating how fast the send-stream should transmit data.
+    context
+        Device context to create buffers.
+    stream_factory
+        Callback function to create the spead2 send stream. It is passed the
+        stream ocnfiguration and memory buffers.
+    packet_payload
+        Size, in bytes, for the output packets (tied array channelised voltage
+        payload only, headers and padding are added to this).
+    tx_enabled
+        Enable/Disable transmission. Defaults to starting enabled.
+    """
+
+    descriptor_heap: spead2.send.Heap
+    header_size: Final[int] = 64
+
+    def __init__(
+        self,
+        output: BOutput,
+        heaps_per_fengine_per_chunk: int,
+        n_tx_items: int,
+        n_channels_per_substream: int,
+        spectra_per_heap: int,
+        timestamp_step: int,
+        send_rate_factor: float,
+        channel_offset: int,
+        context: AbstractContext,
+        stream_factory: Callable[[spead2.send.StreamConfig, Sequence[np.ndarray]], "spead2.send.asyncio.AsyncStream"],
+        packet_payload: int = DEFAULT_PACKET_PAYLOAD_BYTES,
+        tx_enabled: bool = False,
+    ) -> None:
+        # TODO: Update to track individual substreams once multiple beams are supported
+        self.tx_enabled = tx_enabled
+        self.n_beams = 1
+
+        self._chunks_queue: asyncio.Queue[Chunk] = asyncio.Queue()
+        buffers: list[np.ndarray] = []
+
+        # `n_heaps_to_send` is actually used to dictate the amount of buffers (in XSend)
+        # So perhaps we need to change the number of buffers to be range(send_free_queue.maxsize)
+        # n_heaps_to_send = len(buffers) // spectra_per_heap
+
+        send_shape = (heaps_per_fengine_per_chunk, self.n_beams, n_channels_per_substream, spectra_per_heap, COMPLEX)
+        for _ in range(n_tx_items):
+            chunk = Chunk(
+                accel.HostArray(send_shape, SEND_DTYPE, context=context),
+                accel.HostArray(
+                    (heaps_per_fengine_per_chunk, self.n_beams),
+                    np.uint32,
+                ),
+                channel_offset=channel_offset,
+                timestamp_step=timestamp_step,
+            )
+            self._chunks_queue.put_nowait(chunk)
+            buffers.append(chunk.data)
+
+        # Multicast stream parameters
+        self.heap_payload_size_bytes = n_channels_per_substream * spectra_per_heap * COMPLEX * SEND_DTYPE.itemsize
```

Comment:
Do you need this outside the constructor? If not, it doesn't seem to be necessary to store it in the object.